### PR TITLE
`spacetime publish`: Add confirmation for `-c`

### PR DESCRIPTION
**Pull request by bfops**
# Description of Changes

Making this change in response to a request from the BitCraft team.

`spacetime publish -c` now asks for confirmation, unless a new `--force` is passed as well.

# API and ABI breaking changes

This **will break** any automation using `spacetime publish -c`.

I think that's an acceptable level of break? @jdetter ?

# Expected complexity level and risk

*How complicated do you think these changes are? Grade on a scale from 1 to 5,
where 1 is a trivial change, and 5 is a deep-reaching and complex change.*

2

*This complexity rating applies not only to the complexity apparent in the diff,
but also to its interactions with existing and future code.*

*If you answered more than a 2, explain what is complex about the PR,
and what other components it interacts with in potentially concerning ways.*

# Testing
```
Basic publish still works fine, with no input:

  $ echo | spacetime publish -s local test-module
  info: component 'rust-std' for target 'wasm32-unknown-unknown' is up to date
  checking crate with spacetimedb's clippy configuration
      Checking spacetime-module v0.1.0 (/tmp/cramtests-v9blcwxj/publish-test.t/module-test)
      Finished release [optimized] target(s) in 0.15s
      Finished release [optimized] target(s) in 0.06s
  Optimising module with wasm-opt...
  Could not find wasm-opt to optimise the module.
  For best performance install wasm-opt from https://github.com/WebAssembly/binaryen/releases.
  Continuing with unoptimised module.
  Uploading to local => http://127.0.0.1:3000
  Publishing module...
  Updated database with domain: test-module, address: 0d78dc56e0223a61218791b864064ab5


Publishing with `-c` (now) fails without confirmation:

  $ echo | spacetime publish -s local -c test-module
  info: component 'rust-std' for target 'wasm32-unknown-unknown' is up to date
  checking crate with spacetimedb's clippy configuration
      Checking spacetime-module v0.1.0 (/tmp/cramtests-v9blcwxj/publish-test.t/module-test)
      Finished release [optimized] target(s) in 0.16s
      Finished release [optimized] target(s) in 0.06s
  Optimising module with wasm-opt...
  Could not find wasm-opt to optimise the module.
  For best performance install wasm-opt from https://github.com/WebAssembly/binaryen/releases.
  Continuing with unoptimised module.
  Uploading to local => http://127.0.0.1:3000
  This will DESTROY the current test-module module, and ALL corresponding data.
  Are you sure you want to proceed? (y/N) [deleting test-module] Aborting


...and it succeeds when given confirmation:

  $ echo 'y' | spacetime publish -s local -c test-module
  info: component 'rust-std' for target 'wasm32-unknown-unknown' is up to date
  checking crate with spacetimedb's clippy configuration
      Checking spacetime-module v0.1.0 (/tmp/cramtests-v9blcwxj/publish-test.t/module-test)
      Finished release [optimized] target(s) in 0.14s
      Finished release [optimized] target(s) in 0.06s
  Optimising module with wasm-opt...
  Could not find wasm-opt to optimise the module.
  For best performance install wasm-opt from https://github.com/WebAssembly/binaryen/releases.
  Continuing with unoptimised module.
  Uploading to local => http://127.0.0.1:3000
  This will DESTROY the current test-module module, and ALL corresponding data.
  Are you sure you want to proceed? (y/N) [deleting test-module] Publishing module...
  Updated database with domain: test-module, address: 0d78dc56e0223a61218791b864064ab5


A "yes-like" answer is not good enough!

  $ echo 'yessir' | spacetime publish -s local -c test-module
  info: component 'rust-std' for target 'wasm32-unknown-unknown' is up to date
  checking crate with spacetimedb's clippy configuration
      Checking spacetime-module v0.1.0 (/tmp/cramtests-v9blcwxj/publish-test.t/module-test)
      Finished release [optimized] target(s) in 0.15s
      Finished release [optimized] target(s) in 0.06s
  Optimising module with wasm-opt...
  Could not find wasm-opt to optimise the module.
  For best performance install wasm-opt from https://github.com/WebAssembly/binaryen/releases.
  Continuing with unoptimised module.
  Uploading to local => http://127.0.0.1:3000
  This will DESTROY the current test-module module, and ALL corresponding data.
  Are you sure you want to proceed? (y/N) [deleting test-module] Aborting

The clearly-named long flag works:

  $ echo | spacetime publish -s local --destroy-previous test-module
  info: component 'rust-std' for target 'wasm32-unknown-unknown' is up to date
  checking crate with spacetimedb's clippy configuration
      Checking spacetime-module v0.1.0 (/tmp/cramtests-v9blcwxj/publish-test.t/module-test)
      Finished release [optimized] target(s) in 0.15s
      Finished release [optimized] target(s) in 0.06s
  Optimising module with wasm-opt...
  Could not find wasm-opt to optimise the module.
  For best performance install wasm-opt from https://github.com/WebAssembly/binaryen/releases.
  Continuing with unoptimised module.
  Uploading to local => http://127.0.0.1:3000
  This will DESTROY the current test-module module, and ALL corresponding data.
  Are you sure you want to proceed? (y/N) [deleting test-module] Aborting
  $ echo | spacetime publish -s local --destroy-previous --force test-module
  info: component 'rust-std' for target 'wasm32-unknown-unknown' is up to date
  checking crate with spacetimedb's clippy configuration
      Checking spacetime-module v0.1.0 (/tmp/cramtests-v9blcwxj/publish-test.t/module-test)
      Finished release [optimized] target(s) in 0.15s
      Finished release [optimized] target(s) in 0.06s
  Optimising module with wasm-opt...
  Could not find wasm-opt to optimise the module.
  For best performance install wasm-opt from https://github.com/WebAssembly/binaryen/releases.
  Continuing with unoptimised module.
  Uploading to local => http://127.0.0.1:3000
  Skipping confirmation due to --force.
  Publishing module...
  Updated database with domain: test-module, address: 0d78dc56e0223a61218791b864064ab5

The deprecated long flag works like before:

  $ echo | spacetime publish -s local --clear-database test-module
  info: component 'rust-std' for target 'wasm32-unknown-unknown' is up to date
  checking crate with spacetimedb's clippy configuration
      Checking spacetime-module v0.1.0 (/tmp/cramtests-v9blcwxj/publish-test.t/module-test)
      Finished release [optimized] target(s) in 0.14s
      Finished release [optimized] target(s) in 0.06s
  Optimising module with wasm-opt...
  Could not find wasm-opt to optimise the module.
  For best performance install wasm-opt from https://github.com/WebAssembly/binaryen/releases.
  Continuing with unoptimised module.
  Uploading to local => http://127.0.0.1:3000
  Skipping confirmation due to --force.
  Publishing module...
  Updated database with domain: test-module, address: 0d78dc56e0223a61218791b864064ab5
```

### CODE REVIEW

File: crates/cli/src/subcommands/publish.rs [hunk 1]
```diff
@@ -101,6 +102,20 @@ pub fn cli() -> clap::Command {
...
+                .action(SetTrue)
+                .help("DANGEROUS - Proceed with all actions without waiting for user confirmation")
+        )
+        .arg(
+            Arg::new("deprecated_clear_database")
+                .long("clear-database")
+                .requires("name_or_address")
+                .conflicts_with("destroy_previous")
+                .action(SetTrue)
+                .help("DEPRECATED - Use --destroy-previous"),
```
Comment (jdetter) [inline crates/cli/src/subcommands/publish.rs:117]: Why did we add this?

Comment (jdetter) [reply crates/cli/src/subcommands/publish.rs:117]: I see this is a backwards compatible thing - let's just mark this as an API break and merge it for 0.9.

File: crates/cli/src/subcommands/publish.rs [hunk 2]
```diff
@@ -110,7 +125,8 @@ pub async fn exec(mut config: Config, args: &ArgMatches) -> Result<(), anyhow::E
     let name_or_address = args.get_one::<String>("name|address");
     let path_to_project = args.get_one::<PathBuf>("project_path").unwrap();
     let host_type = args.get_one::<String>("host_type").unwrap();
-    let clear_database = args.get_flag("clear_database");
+    let clear_database = args.get_flag("destroy_previous") || args.get_flag("deprecated_clear_database");
```
Comment (jdetter) [inline crates/cli/src/subcommands/publish.rs:128]: Let's just break the API here and not change this

File: crates/cli/src/subcommands/publish.rs [hunk 3]
```diff
@@ -163,6 +175,31 @@ pub async fn exec(mut config: Config, args: &ArgMatches) -> Result<(), anyhow::E
         database_host
     );
 
+    if clear_database {
+        // This is to preserve backwards-compatibility with the old --clear-database.
+        if force || args.get_flag("deprecated_clear_database") {
```
Comment (jdetter) [inline crates/cli/src/subcommands/publish.rs:180]: remove `|| args.get_flag("deprecated_clear_database")`

File: crates/cli/src/subcommands/publish.rs [hunk 4]
```diff
@@ -24,12 +25,12 @@ pub fn cli() -> clap::Command {
                 .help("The type of host that should be for hosting this module"),
         )
         .arg(
-            // TODO(jdetter): Rename this to --delete-tables (clear doesn't really implies the tables are being dropped)
-            Arg::new("clear_database")
-                .long("clear-database")
+            Arg::new("destroy_previous")
```
Comment (jdetter) [inline crates/cli/src/subcommands/publish.rs:28]: let's keep this as `clear-database` - I realized that this was us trying to not break backwards compatibility but let's just break the API in this case.

**Review by jdetter — COMMENTED** [review]
Let's not try to preserve the API here, just break it and we'll merge this as part of 0.9 next week.

File: crates/cli/src/subcommands/publish.rs (hunk 4 above)
Comment (bfops) [reply crates/cli/src/subcommands/publish.rs:28]: I actually renamed this because of the `TODO` above - I'm fine to revert, but I do agree with it in principle!

**Comment by cloutiertyler** [comment]
Is there a reason we're renaming the long version, and not the short version? I don't think the inconsistency is worth it.

What is wrong with `--clear-database`?

**Comment by jdetter** [comment]
> Is there a reason we're renaming the long version, and not the short version? I don't think the inconsistency is worth it.
> 
> What is wrong with `--clear-database`?

The feedback that you gave at the time was that `--clear-database` implies that you're just dropping all of the rows from the tables instead of deleting the tables completely and creating new ones.

File: crates/cli/src/subcommands/publish.rs [hunk 5]
```diff
@@ -24,12 +25,12 @@ pub fn cli() -> clap::Command {
                 .help("The type of host that should be for hosting this module"),
         )
         .arg(
-            // TODO(jdetter): Rename this to --delete-tables (clear doesn't really implies the tables are being dropped)
-            Arg::new("clear_database")
-                .long("clear-database")
+            Arg::new("delete_tables")
+                .long("delete-tables")
                 .short('c')
```
Comment (jdetter) [inline crates/cli/src/subcommands/publish.rs:30]: keeping the `-c` flag here kinda seems like its legacy. I believe we can't use `-d` because we're using that for the debug flag. Maybe we just keep `--clear-database` and drop the TODO, what do you think @bfops @cloutiertyler ?

File: crates/cli/src/subcommands/publish.rs [hunk 6]
```diff
@@ -163,6 +167,30 @@ pub async fn exec(mut config: Config, args: &ArgMatches) -> Result<(), anyhow::E
...
+            std::io::stdin().read_line(&mut input)?;
+            if input.trim().to_lowercase() != "y" && input.trim().to_lowercase() != "yes" {
+                println!("Aborting");
+                return Ok(());
+            }
+        }
+        query_params.push(("clear", "true"));
+    }
+
     eprintln!("Publishing module...");
```
Comment (jdetter) [inline crates/cli/src/subcommands/publish.rs:194]: Even though its not part of this PR I would recommend changing this to `println` as well

File: crates/cli/src/subcommands/publish.rs (hunk 5 above)
Comment (jdetter) [reply crates/cli/src/subcommands/publish.rs:30]: Tyler already responded on this PR - let's just change this back to `--clear-database` and this is good to merge, thanks Zeke 👍

**Review by jdetter — CHANGES_REQUESTED** [review]
Small request and then good to merge, thanks Zeke 👍

**Review by jdetter — APPROVED** [review]
LGTM, thanks Zeke 👍

File: smoketests/__init__.py [hunk 7]
```diff
@@ -129,9 +129,9 @@ def log_records(self, n):
     def publish_module(self, domain=None, *, clear=True, capture_stderr=True):
         publish_output = self.spacetime(
             "publish",
-            *[domain] if domain is not None else [],
             "--project-path", self.project_path,
-            *(["-c"] if clear else []),
+            *[domain] if domain is not None else [],
+            *["-c"] if clear and domain is not None else [],
```
Comment (bfops) [inline smoketests/__init__.py:134]: I just changed `if clear` to `if clear and domain is not None`, but I also moved the lines so the `domain is not None`s are side-by-side.

File: smoketests/__init__.py [hunk 8]
```diff
@@ -130,8 +130,8 @@ def publish_module(self, domain=None, *, clear=True, capture_stderr=True):
         publish_output = self.spacetime(
             "publish",
             *[domain] if domain is not None else [],
+            *["-c"] if clear and domain is not None else [],
             "--project-path", self.project_path,
-            *(["-c"] if clear else []),
```
Comment (bfops) [inline smoketests/__init__.py:134]: (specifying `-c` without a `domain` is now a CLI error)
